src/mlpack/methods/ann/layer/separable_convolution_impl.hpp

```diff
@@ -0,0 +1,495 @@
+      [](unsigned char c){ return std::tolower(c); });
+
+  if (paddingTypeLow == "valid")
+  {
+    padWLeft = 0;
+    padWRight = 0;
+    padHTop = 0;
+    padHBottom = 0;
+  }
+  else if (paddingTypeLow == "same")
+  {
+    InitializeSamePadding();
+  }
+
+  padding = ann::Padding<>(padWLeft, padWRight, padHTop, padHBottom);
+}
+
+template <
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+SeparableConvolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+>::SeparableConvolution(
+    const size_t inSize,
+    const size_t outSize,
+    const size_t kernelWidth,
+    const size_t kernelHeight,
+    const size_t strideWidth,
+    const size_t strideHeight,
+    const std::tuple<size_t, size_t> padW,
+    const std::tuple<size_t, size_t> padH,
+    const size_t inputWidth,
+    const size_t inputHeight,
+    const size_t numGroups,
+    const std::string &paddingType) :
+    inSize(inSize),
+    outSize(outSize),
+    kernelWidth(kernelWidth),
+    kernelHeight(kernelHeight),
+    strideWidth(strideWidth),
+    strideHeight(strideHeight),
+    padWLeft(std::get<0>(padW)),
+    padWRight(std::get<1>(padW)),
+    padHBottom(std::get<1>(padH)),
+    padHTop(std::get<0>(padH)),
+    inputWidth(inputWidth),
+    inputHeight(inputHeight),
+    outputWidth(0),
+    outputHeight(0),
+    numGroups(numGroups)
+{
+  if (inSize % numGroups != 0 || outSize % numGroups != 0)
+  {
+    Log::Fatal << "The output maps / input maps is not possible given "
+        << "the number of groups. Input maps / output maps must be "
+        << " divisible by number of groups." << std::endl;
+  }
+
+  weights.set_size((outSize * (inSize / numGroups) * kernelWidth *
+      kernelHeight) + outSize, 1);
+
+  // Transform paddingType to lowercase.
+  std::string paddingTypeLow = paddingType;
+  std::transform(paddingType.begin(), paddingType.end(), paddingTypeLow.begin(),
+      [](unsigned char c){ return std::tolower(c); });
+
+  if (paddingTypeLow == "valid")
+  {
+    padWLeft = 0;
+    padWRight = 0;
+    padHTop = 0;
+    padHBottom = 0;
+  }
+  else if (paddingTypeLow == "same")
+  {
+    InitializeSamePadding();
+  }
+
+  padding = ann::Padding<>(padWLeft, padWRight, padHTop, padHBottom);
+}
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+void SeparableConvolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+>::Reset()
+{
+  weight = arma::cube(weights.memptr(), kernelWidth, kernelHeight,
+        outSize * (inSize / numGroups), false, false);
+  bias = arma::mat(weights.memptr() + weight.n_elem,
+        outSize, 1, false, false);
```

Comment:
```suggestion
  weight = arma::cube(weights.memptr(), kernelWidth, kernelHeight,
      outSize * (inSize / numGroups), false, false);
  bias = arma::mat(weights.memptr() + weight.n_elem,
      outSize, 1, false, false);
```

Should be tabbed twice (4 spaces).

Reply:
sublime gave me an indent of 2 space and just out of habit i tabbed twice :p , fixed it.